Added hierarchy to hp::FECollection.

marcfehling opened the pull request:
As requested in #7515, we need a hierarchy of finite elements when p refinement or coarsening is executed.

The `hp::FECollection` has been extended by functions that return succeeding and preceding indices compared to the one passed as an argument. By default, a hierarchy depending on the index on the collection is supplied, that can be overriden.

Part of #3511.

bangerth commented on doc/doxygen/headers/hp.h:
```diff
@@ -89,6 +89,13 @@
+ * succeeding finite elements for refinement and preceding ones for coarsening.
+ * By default, we assume that finite elements are stored in an ascending order
+ * based on their polynomial degree. If the order of elements differs,
+ * a corresponding hierarchy needs to be supplied to the hp::FECollection object.
```
Can you provide a link to the function that needs to be called?

bangerth commented on doc/doxygen/headers/hp.h:
```diff
@@ -89,6 +89,13 @@
  *
+ * Whenever p adaptivity is considered in an hp finite element program,
+ * a hierarchy of finite elements needs to be established to determine
+ * succeeding finite elements for refinement and preceding ones for coarsening.
```
```suggestion
 * succeeding finite elements for refinement and preceding ones for coarsening.
 * Typically, this hierarchy considers how finite element spaces are nested:
 * for example, a $Q_1$ element describes a sub-space of a $Q_2$ element,
 * and so doing $p$ refinement usually means using a larger (more accurate)
 * finite element space. In other words, the hierarchy of finite elements is built
 * by considering whether some elements of the collection are sub- or
 * super-spaces of others.
 *
```

bangerth commented on include/deal.II/hp/fe_collection.h:
```diff
@@ -371,6 +372,32 @@ namespace hp
+     */
+    void
+    set_hierarchy(const std::function<unsigned int(const unsigned int)> &next,
+                  const std::function<unsigned int(const unsigned int)> &prev);
```
I would pass the whole `FECollection` as first argument to these functions. Then they can be global functions, for example, that don't need to know about other pieces of the program.

bangerth replied:
There probably also ought to be a `clear_hierarchy()` function. You can then call that from the constructor, because it does what you currently do in the constructor.

bangerth commented on include/deal.II/hp/fe_collection.h:
```diff
@@ -371,6 +372,32 @@ namespace hp
+    /**
+     * Set functions determining the hierarchy of finite elements, i.e. a
+     * function @p next that returns the index of the finite element following
+     * the given one, and a function @p prev returning the preceding one.
```
can you document their arguments?

bangerth commented on include/deal.II/hp/fe_collection.h:
```diff
@@ -371,6 +372,32 @@ namespace hp
+     * @p fe_index in hierarchy.
+     */
+    unsigned int
+    prev_in_hierarchy(const unsigned int fe_index) const;
```
just spell it out -- we like to use whole words in deal.II :-)

bangerth commented on include/deal.II/hp/fe_collection.h:
```diff
@@ -371,6 +372,32 @@ namespace hp
+
+    /**
+     * Function returning the index of the finite element following the given
+     * @p fe_index in hierarchy.
```
I think it would be useful to say what the *default* is and how one can change it.

bangerth commented on source/hp/fe_collection.cc:
```diff
@@ -160,9 +160,25 @@ namespace hp
+    // establish hierarchy corresponding to indices in this collection
+    set_hierarchy(
+      [this](const unsigned int fe_index) {
+        return ((fe_index + 1) < this->size()) ? fe_index + 1 : fe_index;
```
You should document what happens if you're already at the first (last) element of the hierarchy.

bangerth commented on source/hp/fe_collection.cc:
```diff
@@ -202,6 +219,53 @@ namespace hp
+  FECollection<dim, spacedim>::next_in_hierarchy(
+    const unsigned int fe_index) const
+  {
+    Assert(0 <= fe_index && fe_index < size(),
```
here and below, the first half of the condition is implicitly true because you're using an `unsigned` data type

marcfehling commented:
Addressed your suggestions. I'll squash all commits into one once you give your okay.

bangerth commented on include/deal.II/hp/fe_collection.h:
```diff
@@ -54,11 +54,53 @@ namespace hp
+     * succeeding finite elements for refinement and preceding ones for
+     * coarsening.
+     *
+     * In this struct, we supply a hiearchy that is imposed on all FECollection
```
```suggestion
     * In this struct, we supply a hierachy that is imposed on all FECollection
```

bangerth commented on include/deal.II/hp/fe_collection.h:
```diff
@@ -54,11 +54,53 @@ namespace hp
+     * In this struct, we supply a hiearchy that is imposed on all FECollection
+     * objects by default.
+     */
+    struct Hierarchy
```
Would it be useful to call this structure `DefaultHierarchy`?

bangerth commented on include/deal.II/hp/fe_collection.h:
```diff
@@ -371,6 +413,60 @@ namespace hp
+     * Hierarchy::next_index() and Hierarchy::previous_index().
+     */
+    void
+    default_hierarchy();
```
call it `set_default_hierarchy()` to keep the symmetry with the previous function.

bangerth commented:
/rebuild

marcfehling commented:
Ready for merge once tests pass

bangerth commented:
There's a CI failure. Can you look at https://jenkins.tjhei.info/blue/organizations/jenkins/dealii-tidy/detail/PR-7767/9/pipeline ?

marcfehling commented:
Forgot to update syntax changes in the provided test. I fixed that in a commit that still has to be pushed.

However, I tried to reproduce the `clang-tidy` issue on my machine with `clang 7.0.1`, but failed to do so. My log reads:
```
[  0%] Built target obj_sundials_inst
[  0%] Built target expand_instantiations_exe
[ 17%] Built target obj_numerics_inst
[ 41%] Built target obj_fe_inst
[ 47%] Built target obj_dofs_inst
[ 58%] Built target obj_lac_inst
[ 64%] Built target obj_base_inst
[ 70%] Built target obj_gmsh_inst
[ 76%] Built target obj_grid_inst
[ 76%] Built target obj_hp_inst
[ 82%] Built target obj_multigrid_inst
[ 88%] Built target obj_distributed_inst
[ 94%] Built target obj_algorithms_inst
[ 94%] Built target obj_matrix_free_inst
[ 94%] Built target obj_meshworker_inst
[ 94%] Built target obj_opencascade_inst
[ 94%] Built target obj_particle_inst
[100%] Built target obj_differentiation_ad_inst
[100%] Built target obj_physics_inst
[100%] Built target obj_physics_elasticity_inst
[100%] Built target obj_rol_inst
[100%] Built target obj_non_matching_inst
[100%] Built target expand_all_instantiations
OK
```

The CI with `clang 6.0.0` reports:
```
/jenkins/workspace/dealii-tidy_PR-7767/include/deal.II/hp/fe_collection.h:139:5: error: exception specification of explicitly defaulted move constructor does not match the calculated one [clang-diagnostic-error]
```
I could try out `clang 6.0.0` as used by CI to reproduce it on my machine.

I removed the `default` flag of the `FECollection` constructor to register a default hierarchy, but from my understanding this should not affect the move constructor. Or did I get something wrong?

marcfehling commented:
I could not reproduce this issue on my machine with `clang 6.0.0`.
I'll push the overdue commit and hope that this will do the trick.

marcfehling commented:
@tjhei - Does `dealii-tidy` use `clang` with `libc++`?

According to this [stackoverflow entry](https://stackoverflow.com/questions/21320391/can-i-mark-a-classes-move-operation-noexcept-if-it-contains-a-standard-container), it depends on the standard library implementation whether standard containers can be `noexcept`-moved or not. We could get around it via:
```
FECollection(FECollection<dim, spacedim> &&)
  noexcept(std::is_nothrow_move_constructible<std::vector<std::shared_ptr<const FiniteElement<dim, spacedim>>>>::value
           &&
           std::is_nothrow_move_constructible<std::function<unsigned int(const typename hp::FECollection<dim, spacedim> &, const unsigned int)>>::value)
  = default;
```
or do it without the `noexcept` flag.

I'll give it a try.

masterleinad commented:
Can't you just remove the `noexcept` specifier? It should be determined automatically anyway.

marcfehling commented:
`dealii-tidy` accepts the verbose `noexcept` variant.

However, I prefer @masterleinad's suggestion, also for cosmetic reasons :-)

I'll prepare the final version with the upcoming commit.

bangerth commented:
/rebuild

tjhei commented:
> @tjhei - Does `dealii-tidy` use `clang` with `libc++`?

We are running it without libc++. Not sure if there are pro/cons about that.

marcfehling commented:
It looks like `clang-tidy` requires the `noexcept` specifier for the move constructor:
```
/jenkins/workspace/dealii-tidy_PR-7767/include/deal.II/hp/fe_collection.h:139:5: error: move constructors should be marked noexcept [performance-noexcept-move-constructor,-warnings-as-errors]
```
but has problems with certain implementations of standard datatypes, i.e. `std::is_nothrow_move_constructible<T>` is false for some types:
```
/jenkins/workspace/dealii-tidy_PR-7767/include/deal.II/hp/fe_collection.h:139:5: error: exception specification of explicitly defaulted move constructor does not match the calculated one [clang-diagnostic-error]

```

I guess we have to fall back to the (ugly) verbose variant.

bangerth commented:
That's ugly. But we can address this in a follow-up. No need to hold this PR up for it.
